Short-term storage

flomnes opened the pull request:
### TODO

- [x] Check that STS are available also in Adequacy mode -> seems fine, need to test with output files PR #1283 
- [x] Fix units for file "values" #1248 
- [x] Do not try to load STS if the version number is < 860 -> done in #1239
- [x] Output files #1238 
- [ ] Add unit & end-to-end tests / unit tests for loading OK
- [x] Leap year testing / cycles
- [x] Year starting in a month != January / cycles
- [ ] Cycles for days simulations instead of weeks

flomnes committed:
Short-term storage skeleton

flomnes committed:
Rename container class

flomnes committed:
Rename Validate -> validate

flomnes committed:
Fix build

flomnes committed:
Rename Part -> Container

flomnes committed:
Fix #include

flomnes committed:
Add container.cpp

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/short-term-storage

flomnes committed:
Clean up headers, add <vector>

flomnes committed:
Use 'unsigned int'

flomnes committed:
Add SIM data for short-term storage

flomnes committed:
Revert Container::validate

flomnes commented on src/solver/simulation/sim_structure_probleme_economique.h:
```diff
@@ -153,6 +151,35 @@ typedef struct
+    double injectionCapacity;
+    double withdrawalCapacity;
+    double efficiency;
+};
```
Pointer to input data ?

flomnes committed:
Remove nb of units, rename capacity

flomnes committed:
Add missing initial level

flomnes committed:
Add missing global index

flomnes committed:
Add missing NS resolution

flomnes committed:
Rename production -> injection in results

flomnes committed:
Add storagesById, storagesByIndex

flomnes committed:
Allocate / deallocate data for results

flomnes committed:
Allocate / deallocate input data for STS (SIM)

flomnes committed:
Add Runtime::shortTermStorageCount, use pointer-to-vector to avoid segfaults

flomnes committed:
Use pointer-to-vector

flomnes committed:
Allocate/deallocate variable correspondance arrays

flomnes committed:
Revert "Allocate/deallocate variable correspondance arrays"

This reverts commit 2a47cc4bbcdaa186b46d9ae53258f008a9de02dd.

flomnes committed:
Fix segfault, swap indices

flomnes committed:
Fix variable & constraint counting

sylvlecl commented on src/solver/simulation/sim_structure_probleme_economique.h:
```diff
@@ -514,6 +545,8 @@ struct PROBLEME_HEBDO
 
     PALIERS_THERMIQUES** PaliersThermiquesDuPays = nullptr;
     ENERGIES_ET_PUISSANCES_HYDRAULIQUES** CaracteristiquesHydrauliques = nullptr;
+    std::vector<::ShortTermStorage::AREA_INPUT>* ShortTermStorage = nullptr;
```
Is it necessary to use a pointer instead of plain vector?

flomnes replied:
It turns out : no. Not here at least.

It depends on how the superseding structure is allocated. `PROBLEME_HEBDO` objects are allocated using `new`, so all all data members are properly initialized.

On the other hand, `malloc` only allocates memory, it doesn't call constructors for data members. So for example this will fail
```cpp
#include <cassert>
#include <cstdlib>

struct S
{
  int a;
  S() : a(5) {}
};

int main() {
  auto s = static_cast<S*>(malloc(sizeof(S)));
  assert(s->a == 5); // Failure
  free(s);
}
```
Objects of type `RESULTATS_HORAIRES` are allocated using `malloc`, so I got a segfault ;)

sylvlecl commented on src/solver/simulation/sim_structure_probleme_economique.h:
```diff
@@ -433,7 +462,9 @@ typedef struct
-    PRODUCTION_THERMIQUE_OPTIMALE** ProductionThermique;
+    PRODUCTION_THERMIQUE_OPTIMALE** ProductionThermique; // index is pdtHebdo
+
+    std::vector<::ShortTermStorage::RESULTS>* ShortTermStorage = nullptr;
```
Is it necessary to use a pointer instead of plain vector?

flomnes replied:
For this one, yes (see above)

sylvlecl replied:
Understood! this will have to wait for deeper changes :)

sylvlecl commented on src/solver/simulation/sim_structure_probleme_economique.h:
```diff
@@ -153,6 +151,37 @@ typedef struct
+struct RESULTS
+{
+    // Index is the number of the STS in the area
+    std::vector<double> level;      // MWh
```
What about having a small struct for each STS, like in the inputs ?

flomnes replied:
That would be nice, will see if it works with the rest of the code. I essentially re-used the same logic as thermal clusters (multiple clusters per area, input & output data)

sylvlecl replied:
Ok, no strong opinion here: it can be better to keep consistent with existing code, waiting for a larger refactoring in the future

flomnes replied:
It's no big deal, just swap 2 indices

sylvlecl commented on src/libs/antares/study/parts/short-term-storage/unit.h:
```diff
@@ -0,0 +1,45 @@
+    bool loadFromFolder(const std::string& folder);
+
+private:
+    Series series;
```
I guess those fields will need to have some public access

flomnes replied:
@payetvin please consider this

sylvlecl commented on src/libs/antares/study/parts/short-term-storage/container.h:
```diff
@@ -0,0 +1,50 @@
+    bool loadSeriesFromFolder(const std::string& folder);
+
+    std::vector<Unit> storagesByIndex;
+    // TODO[FOM] : fill this map with the address from storagesByIndex
```
It would make sense to have those fields not accessible for modification by outside world, to avoid that kind of consistency issue.

I am not sure of the workflow, but we could have a constructor with only the vector as parameter, which will populate the map, and only expose the vector and the map as const references (or not exposing the map at all and expose a `const Unit* getStorage(const std::string& id)` or something alike).

flomnes replied:
@payetvin please consider this

payetvin committed:
[DEV] Add function to load from section and properties

payetvin committed:
[DEV] Add map for group enum, area loading of sts

payetvin committed:
[DEV] Add logs for group

payetvin committed:
[DEV] Swap pointer between storage map and vector

payetvin committed:
Merge branch 'feature/short-term-storage' into feature/ini-sts

payetvin committed:
[DEV] Fill map before vector

payetvin committed:
[DEV] Add properties.cpp and unit.cpp

payetvin committed:
[DEV] Add validate functions

flomnes committed:
De-allocate memory once when multiple areas are present

payetvin committed:
[DEV] Change names for container and unit

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/short-term-storage

flomnes committed:
Remove useless struct AREA_INPUT

flomnes committed:
Rename SINGLE_OBJECT_INPUT -> PROPERTIES

flomnes committed:
Add property initialLevel

payetvin committed:
[DEV] Add base to load series

payetvin committed:
Merge branch 'feature/short-term-storage' into feature/ini-sts

flomnes committed:
Use std::optional in SIM

payetvin committed:
[DEV] Accept optim arg for initiallevel, change in series construction path

payetvin committed:
[DEV] Comments

payetvin committed:
[DEV] Skip short term storage loading if no folder

payetvin committed:
[DEV] Test for windows error

payetvin committed:
Merge branch 'feature/ini-sts' into feature/short-term-storage

payetvin committed:
[DEV] Test using yuni separator

payetvin committed:
Merge branch 'feature/ini-sts' into feature/short-term-storage

flomnes commented on src/libs/antares/study/area/list.cpp:
```diff
@@ -1279,6 +1287,35 @@ bool AreaList::loadFromFolder(const StudyLoadOptions& options)
+    {
+        logs.info() << "Loading short term storage clusters...";
+        buffer.clear() << pStudy.folderInput << SEP << "st-storage" << SEP << "areas.ini";
+        struct stat sb;
+        if (!stat(buffer.c_str(), &sb))
```
`IO::Directory::Exists`

flomnes replied:
https://en.cppreference.com/w/cpp/filesystem/exists
https://en.cppreference.com/w/cpp/filesystem/directory_entry/is_directory

guilpier-code committed:
Short term storage : hour for clusters functions (#1208)

* [skip ci] st storage output : some cleaning

* st storage output (refactor) : moving code for renewable clusters from top to renewable output variables

* st storage output (refactor) : some cleaning

- Class state : some variables should be local instead of public attributes
- Output variables classes : hourForEachThermalCluster removed when empty

* st storage output (refactor) : output variables related to thermal energy manage their clusters output once for all.

Output variables related to thermal energy manage their clusters once output instead of managing them one by one.

* st storage output (refactor) : remove unnecessary hourForEachThermalCluster(...) and hourForEachRenewableCluster(...) member functions

* st storage output (refactor) : renaming hourForEachThermalCluster(...) into hourForThermalClusters(...)

* st storage output (refactor) : removing unused function

* st storage output (refactor) : renaming hourForThermalClusters(...) into hourForClusters(...)

* st storage output (refactor) : fix a regression in results

The pb occurred when renaming functions hourForThermalClusters : apparently, MSVC did not do the job correctly

* st storage output (refactor) : make test python scripts to print the names of failing output variable

* st storage output (refactor) : test python scripts : correction

* Rename cluster_index -> clusterIndex

* Apply clang-format

* Rename cluster_index -> clusterIndex

* Factorize using lambda function

* Apply clang-format

---------

Co-authored-by: Florian OMNES <florian.omnes@rte-france.com>

flomnes committed:
Fix absence of allocation for PMinOfClusters (#1224)

guilpier-code committed:
ST storage output : adding ST storage output variables (#1217)

* [skip ci] st storage output : some cleaning

* st storage output (refactor) : moving code for renewable clusters from top to renewable output variables

* st storage output (refactor) : some cleaning

- Class state : some variables should be local instead of public attributes
- Output variables classes : hourForEachThermalCluster removed when empty

* st storage output (refactor) : output variables related to thermal energy manage their clusters output once for all.

Output variables related to thermal energy manage their clusters once output instead of managing them one by one.

* st storage output (refactor) : remove unnecessary hourForEachThermalCluster(...) and hourForEachRenewableCluster(...) member functions

* st storage output (refactor) : renaming hourForEachThermalCluster(...) into hourForThermalClusters(...)

* st storage output (refactor) : removing unused function

* st storage output (refactor) : renaming hourForThermalClusters(...) into hourForClusters(...)

* st storage output (refactor) : fix a regression in results

The pb occurred when renaming functions hourForThermalClusters : apparently, MSVC did not do the job correctly

* st storage output (refactor) : make test python scripts to print the names of failing output variable

* st storage output (refactor) : test python scripts : correction

* st storage output : adding a multiple output var for injection and withdrawal in files values-<period>.txt

* st storage output : adding a multiple output var for levels in files values-<period>.txt

* st storage output : adding generation of output for ST storage spatial aggregate (districts)

* Inject simulation results into output variable writer (#1223)

* Inject results (withdrawal/injection)

* Inject results (levels)

* Use unsigned int

* Avoid using temporary maps to store results

---------

Co-authored-by: Florian OMNES <florian.omnes@rte-france.com>

flomnes committed:
Optimization for short term storage (#1201)

* Add correspondance for 3x variables, 1x constraint

* Clean up headers

* Add STS contribution to balance constraint (P=C)

* Build 3x hourly variables for each STS object

* Add bounds for STS variables

* Add address for variables

* Fix variable type

* Use pointer-to-vector

* Bit of cleaning

* Allocate/deallocate constraint correspondance arrays for STS

* Apply clang-format

* Rename shortTermStorage -> shortTermStorageBalance

* Allocate/deallocate variable correspondance arrays

* Add level constraint (inflows = 0)

* Add RHS

* Add missing efficiency factor in level equation

* Fix balance constraint, constraint number

* Move short-term levels constraint into function

* Cycle level constraint

* Fix nb of terms

* Initialize short term storage costs

* Copy STS data from libantares to SIM

* Add public to Properties

* Remove ugly hack

* Fix comp warning

* Use anonymous struct for var correspondance

* Add initial levels

* Use initial level in OPT

* Revert "Merge remote-tracking branch 'github/feature/ini-sts' into feature/optimization-short-term-storage"

This reverts commit 7490a507fa324cfd102c9b448b91d16af9748aa0, reversing
changes made to 548e783d0dca7032414b0b4355e14753457ca597.

* Swap comparisons

* Remove debug log

* Formatting

* Fix merge conflicts (build)

* Revert some changes, fix compilation warnings

* Apply clang-format

* Rename NumberOfShortTermStorage -> NumberOfShortTermStorages

* Add forgotten study/parts/short-term-storage/cluster.h

payetvin committed:
[DEV] Fix folder not found

payetvin committed:
[DEV] Code smells

flomnes committed:
Set group 1 as the default

flomnes commented on src/solver/variable/economy/shortTermStorageLevels.h:
```diff
@@ -0,0 +1,267 @@
+            const auto cluster = state.area->shortTermStorage.storagesByIndex[stsIndex];
+            const uint group = Antares::Data::ShortTermStorage::groupIndex(cluster->properties.group);
+            pValuesForTheCurrentYear[numSpace][group][state.hourInTheYear] += (*state.hourlyResults->ShortTermStorage)[state.hourInTheWeek].level[stsIndex];
+        }
```
This is where the extraction takes place

flomnes committed:
Fix code smells

flomnes committed:
Apply clang-format

sylvlecl commented on src/libs/antares/study/parts/short-term-storage/cluster.h:
```diff
@@ -0,0 +1,50 @@
+
+namespace Antares::Data::ShortTermStorage
+{
+class STstorageCluster
```
capital S if we want to respect `CamelCase`

sylvlecl commented on src/libs/antares/study/parts/short-term-storage/properties.h:
```diff
@@ -0,0 +1,77 @@
+{
+public:
+    bool validate();
+    bool loadKey(const IniFile::Property* p);
```
From the implementation it seems the property is expected to be not-null, we must use a reference instead of a pointer

payetvin replied:
if (!section.firstProperty)
        return false;
We check if first property isn't null and then iterate till null so no need to check in that function

sylvlecl replied:
Yes, I am not saying that there is an error in the current code, but the point is to make clear to the caller of this method, who could possibly be anyone in the future since it's public, that he must not pass a null argument.

Using a reference is a semantic way in c++ to document that an argument is not nullable.

sylvlecl commented on src/libs/antares/study/parts/short-term-storage/properties.h:
```diff
@@ -0,0 +1,77 @@
+    // cluster name
+    std::string name;
+
+    static const std::map<std::string, enum Group> stStoragePropertyGroupEnum;
```
We have not defined naming conventions, but I would suggest to name it with UPPER_CASE to make clear it's a constant.

sylvlecl commented on src/libs/antares/study/parts/short-term-storage/properties.cpp:
```diff
@@ -0,0 +1,147 @@
+    case Group::Other4: return 7;
+    case Group::Other5: return 8;
+    default:
+      return 0;
```
Shouldn't we throw an error here ?
If we add new values some day and forget to update this, it may cause silent bugs.

payetvin replied:
We return an unsigned int so I don't know what value to return for error

flomnes replied:
[`throw`](https://en.cppreference.com/w/cpp/language/throw)

sylvlecl commented on src/libs/antares/study/parts/short-term-storage/properties.h:
```diff
@@ -0,0 +1,77 @@
+*/
+
+#include <optional>
+#pragma once
```
Should be above include?

sylvlecl commented on src/libs/antares/study/parts/short-term-storage/properties.h:
```diff
@@ -0,0 +1,77 @@
+    std::optional<double> initialLevel;
+    // Efficiency factor between 0 and 1
+    double efficiencyFactor;
+    // Cycle duration, 1 <= cycleDuration <= 168
```
This is not checked at validation time

payetvin replied:
TODO
Need to get Simplex optimization range global parameter to validate cycle duration

sylvlecl commented on src/solver/variable/state.h:
```diff
@@ -169,24 +162,20 @@ class State
-    //! The operating cost for the current cluster of the current hour (production level*production
+
+    //! Thermal production for thermal clusters for the current hour in the year
+    std::vector<std::vector<double>> thermalClustersProductions;
```
This refactoring really does not clarify the class from my point of view:
1. why do we have data for all clusters and areas, whereas we have a "current cluster" and a "current area" ?
2. what is the indexing of those vectors ? In general it  is better to create strucs/classes instead of having nested collections, or at least expose getters so that users of the class don't HAVE to know that implementation detail

flomnes replied:
See #1230

sylvlecl commented on src/solver/optimisation/opt_construction_matrice_des_contraintes_cas_lineaire.cpp:
```diff
@@ -73,6 +70,95 @@ void exportPaliers(const PROBLEME_HEBDO& problemeHebdo,
+        int nombreDeTermes = 0;
+        const int globalIndex = storage.globalIndex;
+        // L[h+1] - L[h] - efficiency * injection[h] + withdrawal[h] = inflows[h]
+        if (const int varLevel_next = VarOptim_next->ShortTermStorage.LevelVariable[globalIndex];
```
For the future, I think we should think of syntactic improvements to make this kind of repetitive code more concise, readable and maintainable (always the same operations: get var number, fill vectors, increment count ... ).
For ex:
```cpp
// Helper class to build constraints
class ConstraintFiller {
public:
    ConstraintFiller(double *pi, int *colonne, int &nombreDeTermes)
            : Pi(pi), Colonne(colonne), nombreDeTermes(nombreDeTermes) {}

    void addTerm(int numVar, double value) {
        if (numVar >= 0)
        {
            Pi[nombreDeTermes] = value;
            Colonne[nombreDeTermes] = numVar;
            nombreDeTermes++;
        }
    }

private:
    double* Pi;
    int* Colonne;
    int& nombreDeTermes;
};


// Then in the method only one-liners
// L[h+1] - L[h] - efficiency * injection[h] + withdrawal[h] = inflows[h]
auto& nextHourStorage = VarOptim_next->ShortTermStorage;
auto& currentStorage = VarOptim_current->ShortTermStorage;
ConstraintFiller filler(Pi, Colonne, nombreDeTermes);
filler.addTerm(nextHourStorage.LevelVariable[globalIndex], 1.0);
filler.addTerm(currentStorage.LevelVariable[globalIndex], -1.0);
filler.addTerm(currentStorage.InjectionVariable[globalIndex], -1.0 * storage.efficiency);
filler.addTerm(currentStorage.WithdrawalVariable[globalIndex], 1.0);
```

sylvlecl replied:
(or even better with method chaining, to avoid naming and repeating the "filler" variable)

flomnes replied:
See #1234

flomnes commented:
Careful: name != ID.

IDs are "sanitized" using function `TransformNameIntoID`, making sure all characters are "simple" and can be contained in a filename.

payetvin committed:
Loading Series for short term storage (#1228)

* [DEV] Base for series

* [DEV] Series working, need improvement

* [DEV] Small fixes

* [FIX] Code smell reference

* [DEV] Comments and improvements

* [DEV] Remove logs and comments

* [DEV] Add default values and validate for series

* [DEV] Comments

* [DEV] Add param size to Series::loadFile

* [DEV] Use function in antares/lib to read series files

* [DEV] Error management when loading series

* [BUG] Not enough memory, trying to fix

* Move some code to Series

* Apply clang-format

* Remove logs

* Rename inflow.txt -> inflows.txt

* [DEV] Add debug logs

* [DEV] Try using std::optional for properties

* Call std::optional::value, fix build

* Clarify mandatory properties

* [DEV] Add lambda

* Small adjustments

* Capture + clang-format

* Fix default value for inflows

---------

Co-authored-by: Florian OMNES <florian.omnes@rte-france.com>

flomnes committed:
Use series in optimization (#1235)

* [DEV] Base for series

* [DEV] Series working, need improvement

* [DEV] Small fixes

* [FIX] Code smell reference

* [DEV] Comments and improvements

* [DEV] Remove logs and comments

* [DEV] Add default values and validate for series

* [DEV] Comments

* [DEV] Add param size to Series::loadFile

* [DEV] Use function in antares/lib to read series files

* [DEV] Error management when loading series

* [BUG] Not enough memory, trying to fix

* Move some code to Series

* Apply clang-format

* Remove logs

* Rename inflow.txt -> inflows.txt

* [DEV] Add debug logs

* Use inflows in optimization problem

* Use modulation & rule curves

* Avoid copy/move

* [DEV] Try using std::optional for properties

* Call std::optional::value, fix build

* Clarify mandatory properties

* [DEV] Add lambda

* Small adjustments

* Capture + clang-format

* Fix default value for inflows

* Add & use STStorageInput::count

---------

Co-authored-by: Vincent Payet <vincent.payet@rte-france.com>

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/short-term-storage

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/short-term-storage

flomnes committed:
Use `shared_ptr` instead of raw pointer (#1237)

* Use std::vector + std::move, not raw pointer

* Use std::shared_ptr<Series>

flomnes committed:
Use HOURS_PER_YEAR from antares/constants.h (#1240)

payetvin committed:
Code smells for short term storage (#1242)

* [FIX] Code smell #1

* [FIX] Code smell #2

payetvin commented:
Move `bool loadFile(const std::string& path, std::vector<double>& vect)` in series.cpp to an I/O file

payetvin committed:
Unit testing for short term storage loading (#1236)

* [TESTS] Start unit testing for short term storage

* [TESTS] Add testing for loading series files

* [TESTS] Add testing for loading and filling series

* [TESTS] Add cluster series testing

* [TESTS] Fix ctest command and add properties testing

* [TESTS] Add properties testing

* [TESTS] Add function getFolder

* [TESTS] remove log and add todo

* [TESTS] Fix array1d

* [DEV] Improve validate series

* [DEV] Add check for some values

* [DEV] Add random tests

* [TESTS] Try fixing windows file issues

* [TESTS] Fix cluster.series shared pointer

* [TESTS] Add log for windows

* [FIX] Using own function to read series file

* [FIX] loadFile and loadVector now outside of series class

* [DEV] Use HOURS_PER_YEAR in loadFile

* [DEV] Clean code and logs

* [DEV] Clean and comment for series

* [DEV] Add getFolder and removeFileSeries to fixture

* [TESTS] Fix log and catch exception

* [TESTS] Remove unused includes

* [TESTS] Remove usleess try catch

flomnes committed:
Refactor thermal state (#1230)

* Refactor thermal state

* Move ThermalState::initializeFromArea def to state.cpp

* Fix segfault

* Avoid re-allocating memory too often

* Merge hourForClusters into hourForEachArea

* Remove comment

* Add `explicit` keyword to `ThermalState`'s constructor

* PR remarks

guilpier-code commented on src/libs/antares/study/area/list.cpp:
```diff
@@ -1279,6 +1290,28 @@ bool AreaList::loadFromFolder(const StudyLoadOptions& options)
+
+            for (const auto& [id, area] : areas)
+            {
+                buffer.clear() << pStudy.folderInput << stStoragePlant << area->id;
```
About : 
```C++
buffer.clear() << pStudy.folderInput << stStoragePlant << area->id;
```
We don't need to clear the buffer, it already contains a useful piece of address from line : 
```C++
buffer.clear() << pStudy.folderInput << SEP << "st-storage";
```
above.
Just add the "clusters" part.

payetvin replied:
done in #1246

guilpier-code commented on src/libs/antares/study/parts/short-term-storage/properties.h:
```diff
@@ -0,0 +1,76 @@
+    bool validate();
+    bool loadKey(const IniFile::Property* p);
+    // Not optional   Injection nominal capacity, >= 0
+    std::optional<double> injectionCapacity;
```
Names do not fit the specifications :  
- **injectionCapacity** should be renamed **injectionNominalCapacity**, the **nominal** missing piece being quite meaningful.
- Same for withdrawal
- **capacity** should renamed **reservoirCapacity**

flomnes replied:
@payetvin Can you perform this renaming ?

payetvin replied:
should be done after merging #1239

guilpier-code commented on src/solver/simulation/sim_structure_probleme_economique.h:
```diff
@@ -153,6 +159,32 @@ typedef struct
+
+    std::shared_ptr<Antares::Data::ShortTermStorage::Series> series;
+
+    int globalIndex;
```
`int globalIndex` :
I wondered for a short moment what exactly this index stood for.
If it was called **clusterGlobalIndex** (for instance), I would not have.

flomnes replied:
@payetvin Can you perform this renaming ?

payetvin committed:
Validation of inputs for sts (#1246)

* [DEV] Switch to checkBetween to validate PMAX

* [DEV] add validation for inflows and rule curves

* [DEV] remove validation for inflows, need cycle

* [DEV] remove ruleCurveValidate

* [DEV] add validate function for each vector

* [DEV] fix validate series

guilpier-code commented on src/libs/antares/study/area/list.cpp:
```diff
@@ -1079,12 +1079,23 @@ static bool AreaListLoadFromFolderSingleArea(Study& study,
         buffer.clear() << study.folderInput << SEP << "renewables" << SEP << "series";
         ret = area.renewable.list.loadDataSeriesFromFolder(study, options, buffer) && ret;
     }
+    if (!ret)
```
```C++
if (!ret)
   logs.error() << "Error while loading sts series";
```
Why not right after the ST storage load ? **ret == false** here could mean there were a problem when loading renewable clusters, not when loading ST storage time series.

payetvin replied:
Changed in #1239

guilpier-code commented on src/libs/antares/study/runtime/runtime.h:
```diff
@@ -171,6 +171,8 @@ class StudyRuntimeInfos
     uint thermalPlantTotalCount;
     uint thermalPlantTotalCountMustRun;
 
+    uint shortTermStorageCount;
```
Why not : 
```C++
unsigned int shortTermStorageCount = 0;
```
?

flomnes replied:
See 4187a85

guilpier-code commented on src/libs/antares/study/runtime/runtime.cpp:
```diff
@@ -409,6 +411,7 @@ StudyRuntimeInfos::StudyRuntimeInfos(uint nbYearsParallel) :
  bindingConstraint(nullptr),
  thermalPlantTotalCount(0),
  thermalPlantTotalCountMustRun(0),
+ shortTermStorageCount(0),
```
If `shortTermStorageCount` already initialized in **runtime.h**, we don't need to initialize it here.

guilpier-code commented on src/libs/antares/study/runtime/runtime.cpp:
```diff
@@ -446,7 +449,7 @@ bool StudyRuntimeInfos::loadFromStudy(Study& study)
     thermalPlantTotalCount = 0;
     thermalPlantTotalCountMustRun = 0;
-
+    shortTermStorageCount = 0;
```
Do we need this initialization (as `shortTermStorageCount` is supposed to be already initialized to **0**) ?

flomnes replied:
See 3910a5e

guilpier-code commented on src/solver/simulation/sim_calcul_economique.cpp:
```diff
@@ -39,6 +39,34 @@ using namespace Antares;
 using namespace Antares::Data;
 using namespace Yuni;
 
+static void importShortTermStorages(
```
We could take advantage of this function to compute the total number of ST storage clusters and store it into `problem.NumberOfShortTermStorages`
We would have : 
```C++
static void importShortTermStorages(const AreaList& areas, PROBLEME_HEBDO& problem)
{
    ...
}
```
Doing this would allow to remove the computation of this number from **runtime.cpp**, and even remove all references to ST storage from class **StudyRuntimeInfos**, which I don't like because it stores all kinds of things, don't store proper runtime data and is probably useless eventually.

If we decide to make this move, we'll have to call function **importShortTermStorages** before function **SIM_AllocationProblemeHebdo** : 
```C++
void SIM_InitialisationProblemeHebdo(Data::Study& study,
                                     PROBLEME_HEBDO& problem,
                                     int NombreDePasDeTemps,
                                     uint numSpace)
{
    ...
    importShortTermStorages(study.areas, problem);
    SIM_AllocationProblemeHebdo(problem, NombreDePasDeTemps);
    ...
}
```

flomnes replied:
I agree, but having `shortTermStorageCount` in runtime.h allows to display this log (runtime.cpp)
```cpp
    logs.info() << "     short-term storages: " << shortTermStorageCount;
```

I'm not sure we could display it so easily if it was only in `PROBLEME_HEBDO`.

guilpier-code commented on src/libs/antares/study/runtime/runtime.h:
```diff
@@ -171,6 +171,8 @@ class StudyRuntimeInfos
     uint thermalPlantTotalCount;
     uint thermalPlantTotalCountMustRun;
 
+    uint shortTermStorageCount;
```
Please see [this comment](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1163#discussion_r1162733128)

It suggests to remove `unsigned int shortTermStorageCount` from **runtime.h**

guilpier-code commented on src/solver/simulation/sim_structure_probleme_economique.h:
```diff
@@ -66,6 +65,12 @@ typedef struct
+        int* InjectionVariable;
+        int* WithdrawalVariable;
+        int* LevelVariable;
+    } ShortTermStorage;
```
There are too many things called **ShortTermStorage** : a namespace, a **struct**, a vector, ...
This is confusing, please try to do some renaming : **st_storage_names**,  ...

flomnes replied:
`SIM_ShortTermStorage` ?

guilpier-code commented on src/solver/optimisation/opt_decompte_variables_et_contraintes.cpp:
```diff
@@ -220,6 +220,15 @@ int OPT_DecompteDesVariablesEtDesContraintesDuProblemeAOptimiser(PROBLEME_HEBDO*
         }
     }
 
+    for (int areaIndex = 0; areaIndex < problemeHebdo->NombreDePays; areaIndex++)
```
Instead of making a loop over areas and pick the number of ST storage clusters in each area, couldn't we use `problemeHebdo->NumberOfShortTermStorages`  ?

flomnes replied:
See 996ae317

guilpier-code commented on src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp:
```diff
@@ -157,6 +157,63 @@ void setBoundsForUnsuppliedEnergy(PROBLEME_HEBDO* problemeHebdo,
+                // 3. Levels
+                int varLevel
+                  = CorrespondanceVarNativesVarOptim->ShortTermStorage.LevelVariable[globalIndex];
+                if (pdtHebdo == PremierPdtDeLIntervalle && storage.initialLevel.has_value())
```
About : 
```C++
if (pdtHebdo == PremierPdtDeLIntervalle && storage.initialLevel.has_value())
```
if the simulation period initial level is fixed by user, why do we need bounds for it ?

flomnes replied:
We could move it to the RHS, but it would force us to make a special case, remove one variable from the problem, etc.

It's possible, but adds a bit of complexity. The decision is to let the solver handle this complexity.

guilpier-code commented on src/libs/antares/study/parts/short-term-storage/container.h:
```diff
@@ -0,0 +1,49 @@
+    // Number of ST storages
+    std::size_t count() const;
+
+    std::vector<STStorageCluster*> storagesByIndex;
```
Why is it called **storagesByIndex** ?  it's simply a list of clusters. Why not call it **clusters** ?

flomnes replied:
To make it clear that variables should be accessed by index. Same idea as `areas.byIndex`

payetvin committed:
[MERGE] develop and fix conflict in state.cpp and sim_structure_probleme_economique.h

payetvin committed:
[FIX] Compile error

flomnes committed:
Introduce weekly checks after loading the calendar (#1249)

* Extract method DayInterval::numberOfWeeks

* Define nbHoursInAWeek only once

* Add checkWeekData entry point for each area

* Reduce number of args, rename, add log message

guilpier-code committed:
ST storage : fix a bug on adq-patch tests due to a previous merge

guilpier-code committed:
storage storage output : files details + values (#1238)

* max nb columns for print : simplify class AllVariablesPrintInfo

* max nb columns for print : simplify class AllVariablesPrintInfo (part 2)

* max nb columns for print : move all computations in one place (SurveyResults's constructor)

* max nb columns for print : computing an accurate max number of columns (based on synthesis reports)

Note that synthesis reports contain more columns than year by year reports.
So the computed max number of columns is actually the max number of columns in a synthesis report

* max nb columns for print : more cleaning (#1180)

* max nb columns for print : code improvement

Improvements :
- At reading the output variable print status (parameters.cpp)
- Small changes inside classes of variable-print-info.h/.cpp
- Removing function SurveyResults::initializeMaxVariables(...), which has shrunk until it does one function call.

* max nb columns for print : in class AllVariablesPrintInfo, changing the std::vector into a std::map

* max nb columns for print : renaming in output variable BC MARG COST

* max nb columns for print : fix a bug

* max nb columns for print : fix other bugs

* max nb columns for print : refresh names lists each time a print status is changed

* max nb columns for print : removing useless isEmpty()

Indeed, allVarsPrintInfo can never be empty, has it is build from code built at compile time

* max nb columns for print : try fix a compilation failure on Linux

* max nb columns for print : change for more readable manner to traverse std::map

* max nb columns for print : corrections after review

* max nb columns for print : more corrections after review

* max nb columns for print : amend previous commit (causes compilation error on Linux)

* max nb columns for print : remove lists that have to be updated in real-time

* Use enum value in template param (#1193)

* max nb columns for print : adding member function numberOfEnabledVariables() to class AllVariablesPrintInfo

---------

Co-authored-by: Florian Omnès <florian.omnes@rte-france.com>

* max nb columns for print : corrections after review

* max nb columns for print : removing unnecessary member functions

* ST storage output : adding short term storage injection by cluster

* ST storage output : adding a member function getMaxNumberColumns() to class STstorageInjectionByCluster

* ST storage output : correction after tests

* ST storage output : correction in synthesis ST storage details files

* ST storage output : merging output variable classes driving value-<period>.txt for injection + withdrawal and level

* ST storage output : adding output variable classes inserting withdrawal and levels in details-STstorage-<period>.txt

* ST storage output : corrections after review

* ST storage output : refactoring class ShortTermStorageByGroup's VCard Caption() function.

* ST storage output : we don't have to multiply levels by reservoir capacity

levels are already expressed in MWh at this point.

* ST storage output : corrections after review

* ST storage output : more corrections after review

* ST storage output : more corrections after review

---------

Co-authored-by: Florian Omnès <florian.omnes@rte-france.com>

flomnes committed:
Simplify computation of variables & constraints

flomnes committed:
Initialize shortTermStorageCount

flomnes committed:
Remove init

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/short-term-storage

guilpier-code committed:
St storage output : adapt Unit() to Multiple output variables (#1248)

* max nb columns for print : simplify class AllVariablesPrintInfo

* max nb columns for print : simplify class AllVariablesPrintInfo (part 2)

* max nb columns for print : move all computations in one place (SurveyResults's constructor)

* max nb columns for print : computing an accurate max number of columns (based on synthesis reports)

Note that synthesis reports contain more columns than year by year reports.
So the computed max number of columns is actually the max number of columns in a synthesis report

* max nb columns for print : more cleaning (#1180)

* max nb columns for print : code improvement

Improvements :
- At reading the output variable print status (parameters.cpp)
- Small changes inside classes of variable-print-info.h/.cpp
- Removing function SurveyResults::initializeMaxVariables(...), which has shrunk until it does one function call.

* max nb columns for print : in class AllVariablesPrintInfo, changing the std::vector into a std::map

* max nb columns for print : renaming in output variable BC MARG COST

* max nb columns for print : fix a bug

* max nb columns for print : fix other bugs

* max nb columns for print : refresh names lists each time a print status is changed

* max nb columns for print : removing useless isEmpty()

Indeed, allVarsPrintInfo can never be empty, has it is build from code built at compile time

* max nb columns for print : try fix a compilation failure on Linux

* max nb columns for print : change for more readable manner to traverse std::map

* max nb columns for print : corrections after review

* max nb columns for print : more corrections after review

* max nb columns for print : amend previous commit (causes compilation error on Linux)

* max nb columns for print : remove lists that have to be updated in real-time

* Use enum value in template param (#1193)

* max nb columns for print : adding member function numberOfEnabledVariables() to class AllVariablesPrintInfo

---------

Co-authored-by: Florian Omnès <florian.omnes@rte-france.com>

* max nb columns for print : corrections after review

* max nb columns for print : removing unnecessary member functions

* ST storage output : adding short term storage injection by cluster

* ST storage output : adding a member function getMaxNumberColumns() to class STstorageInjectionByCluster

* ST storage output : correction after tests

* ST storage output : correction in synthesis ST storage details files

* ST storage output : merging output variable classes driving value-<period>.txt for injection + withdrawal and level

* ST storage output : adding output variable classes inserting withdrawal and levels in details-STstorage-<period>.txt

* ST storage output : corrections after review

* ST storage output : refactoring class ShortTermStorageByGroup's VCard Caption() function.

* ST storage output : we don't have to multiply levels by reservoir capacity

levels are already expressed in MWh at this point.

* ST storage output : corrections after review

* ST storage output : more corrections after review

* ST storage output : more corrections after review

* [skip ci] ST storage output : adapting Unit() static function for Multiple output variables

* ST storage output - adapting Unit() : small correction

* Fix warnings

* clang-format

* Fix Unit

* Add comments

* Return "" instead of NULL

* Use `std::string` for function `Description` (#1258)

* Use std::string for function Description

* Use unsigned int

---------

Co-authored-by: Florian Omnès <florian.omnes@rte-france.com>

flomnes committed:
Remove pointer on `std::vector`s (#1269)

* [STS] Remove pointer on std::vectors

* RESULTATS_HORAIRES** -> std::vector<RESULTATS_HORAIRES>

* Remove obsolete comments

* Improve comments a bit

flomnes committed:
Load STS data only if version >= 860 (#1281)

* Load STS data only if version >= 860

* Load STS properties only if version >= 860

flomnes committed:
Make the initial level a dimensionless quantity, multiply by capacity (#1280)

guilpier-code committed:
Fix indices in modeler (#1267)

* st storage - fix modeler : first try

* [skip ci] st storage - fix modeler : small corrections

* Rename STindex -> storageIndex

* Use count()

---------

Co-authored-by: Florian OMNES <florian.omnes@rte-france.com>

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/short-term-storage

flomnes committed:
Add output variables for short-term storage in adequacy mode (#1283)

* Add STS variables for adequacy study mode

* Apply clang-format on every all.h

payetvin committed:
[FIX] InitialLevel < 1 (#1286)

flomnes committed:
Revert weekly checks (#1285)

flomnes committed:
Fix path when more than 1 ST storage is present in an area

payetvin committed:
[FIX] Renaming
